### ISSUE TITLE
1.8.0 fails to build native extensions with --jobs>1

### ISSUE THREAD

**Issue by antifuchs** [bundler/bundler#3385]
I'm just trying out the new 1.8.0 release, and it's failing to install several gems in our gemfile correctly when used with gems that have native extensions. Versions up to 1.7.13 didn't have that problem. I'm testing this on Linux, with Ruby 2.1.3.

One failing gem is a amatch (redacted output since some parts are sensitive):

The command I ran is: `bundle install --jobs=8 --disable-shared-gems`.

```
/usr/lib/rbenv/versions/2.1/lib/ruby/2.1.0/rubygems/ext/builder.rb:160: warning: conflicting chdir during another chdir block

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    /usr/lib/rbenv/versions/2.1/bin/ruby extconf.rb
creating Makefile
Rewriting the generated Makefile to get my compiler output displayed...

Makefile not found

Gem files will remain installed in /home/vagrant/foo/vendor/bundle/gems/amatch-0.2.11 for inspection.
Results logged to /home/vagrant/foo/vendor/bundle/extensions/x86_64-linux/2.1.0/amatch-0.2.11/gem_make.out
```

(the `gem_make.out` file contains only the "Makefile not found" message above)

Parallel bundler 1.8 is spewing that "conflicting chdir during another chdir block" message a lot - almost every second line of output is that message (1.7.x is completely quiet). That there is no Makefile in the directory left behind seems somewhat related, too (:

Bundling in parallel with 1.8.0 reliably fails; however, unsetting the `BUNDLE_JOBS` variable and bundling in serial always works on both 1.7 and 1.8.

Let me know if you need any more information - happy to run any experiment to track this issue down!

**Comment by Who828** [comment]
Just a couple of questions,

1) How is `--disable-shared-gems' working for you? We removed that flag in 1.1.pre IIRC. Are you using a custom version of Bundler or something? If so just use the default one and see if the issue persists.

2) Can you give us a sample Gemfile to reproduce the issue? I tried creating one at my end but couldn't reproduce it, again it might have to do something with 1.

**Comment by domcleal** [comment]
We see the same errors with Foreman builds under Bundler 1.8.0 and ruby 1.9.3p392 (2013-02-22 revision 39386) [x86_64-linux].

Bundler output is at https://gist.github.com/domcleal/659f94ac68b03c911ccb or http://ci.theforeman.org/job/test_develop_pr_core/3895/database=postgresql,ruby=1.9.3/console.  The command we're running is `bundle install --without=development -j5`.

Our Gemfile is available at https://github.com/theforeman/foreman, just copy the example config/settings.yaml and config/database.yml files after checking it out.

I haven't been able to reproduce it reliably outside of our CI environment unfortunately, which seems to hit it over 50% of the time since 1.8.0.

**Comment by TimMoore** [comment]
@antifuchs / @domcleal could you put the output from `bundle env` in a gist and post a link here please? I'm mostly interested in the Rubygems version, but all of that information is useful.

Also, see if upgrading Rubygems helps (`gem update --system`). There have been some thread-safety issues fixed in the gem installer (https://github.com/rubygems/rubygems/pull/1005) and as @Who828 mentioned, Bundler 1.8 switches parallel installation on MRI from using forked processes to using threads.

**Comment by indirect** [comment]
We might need to abort if jobs is > 1 and rubygems is older than contains that fix.

**Comment by domcleal** [comment]
Thanks for the hint, it seems that the Rubygems version is the cause.  Our CI slaves had mismatched versions, and so it was failing on some slaves and not others.

A working slave has Rubygems 2.2.2 (https://gist.github.com/domcleal/9791f3f62e0b0592c3bc) and a failing slave had Rubygems 1.8.25 (https://gist.github.com/domcleal/f051c69acdadb21ad09e).

Interesting that neither slave has the thread safety fix mentioned above - perhaps there was an earlier one between 1.8.25 and 2.2.2 that's also required.  Aborting when using *some* older version sounds like a good idea.

**Comment by indirect** [comment]
Did some digging with @antifuchs today, and it's an issue where the git gem native extension building code has a race condition with the Rubygems native extension building mutex. Looks like we're going to have to patch Rubygems to fix it for good. :/

**Comment by indirect** [comment]
This should be fixed by cc99df8c426787b8b32ba5c915ff1218243813a9. @antifuchs can you test with [this bundler gem](http://files.arko.net/2i3r3r2C2B3x/bundler-1.8.1.pre.gem) to make sure it works for you?

**Comment by antifuchs** [comment]
@indirect - I just bundled successfully with `--jobs=8` three times in a row; everything totally worked. Pretty glad this works now (:

**Comment by indirect** [comment]
woo, thanks for testing :)

**Comment by indirect** [comment]
Released 1.8.1 with this fix, should be good for everyone now.
